daffodil-runtime2/src/main/resources/org/apache/daffodil/runtime2/c/libruntime/unparsers.c

```diff
@@ -17,135 +17,443 @@
-    }
+    // Fill and copy the fractional byte as many times as needed
+    while (num_bits + ustate->unwritLen >= BYTE_WIDTH)
+    {
+        // Fill the fractional byte
+        uint8_t whole_byte = bytes[ix_bytes++];
+        size_t  num_bits_fill = BYTE_WIDTH - ustate->unwritLen;
+        ustate->bitPos0b += num_bits_fill;
+        ustate->unwritBits <<= num_bits_fill;
+        ustate->unwritBits |= HIGH_BITS(whole_byte, num_bits_fill);
+        ustate->unwritLen += num_bits_fill;
+        num_bits -= num_bits_fill;
+        whole_byte <<= num_bits_fill;
+
+        // Copy the fractional byte to stream
+        size_t num_bits_write = BYTE_WIDTH;
+        size_t count = fwrite(&ustate->unwritBits, 1, 1, ustate->stream);
+        ustate->unwritLen -= num_bits_write;
+        if (count < 1)
+        {
+            ustate->error = eof_or_error(ustate->stream);
+            return;
+        }
 
-#define define_unparse_endian_real(endian, type, bits)                                                       \
-    void unparse_##endian##_##type(type number, UState *ustate)                                              \
-    {                                                                                                        \
-        union                                                                                                \
-        {                                                                                                    \
-            char           c_val[sizeof(type)];                                                              \
-            type           f_val;                                                                            \
-            uint##bits##_t i_val;                                                                            \
-        } buffer;                                                                                            \
-                                                                                                             \
-        buffer.f_val = number;                                                                               \
-        buffer.i_val = hto##endian##bits(buffer.i_val);                                                      \
-        write_stream_update_position(buffer.c_val, sizeof(type));                                            \
+        // Copy any remaining unused bits from the whole byte to the
+        // fractional byte
+        size_t num_bits_unused = BYTE_WIDTH - num_bits_fill;
+        if (num_bits_unused > num_bits) num_bits_unused = num_bits;
+        if (num_bits_unused)
+        {
+            ustate->bitPos0b += num_bits_unused;
+            ustate->unwritBits <<= num_bits_unused;
+            ustate->unwritBits |= HIGH_BITS(whole_byte, num_bits_unused);
+            ustate->unwritLen += num_bits_unused;
+            num_bits -= num_bits_unused;
+        }
```

Comment:
I think I have the same questions for uparse as parse related to bitPos0b and when it's updated.